spectral_cube/lower_dimensional_structures.py

```diff
@@ -1058,41 +1058,136 @@ def _new_spectrum_with(self, data=None, wcs=None, mask=None, meta=None,
+                                unit=unit, fill_value=fill_value,
+                                header=header or self._header,
+                                wcs_tolerance=wcs_tolerance or self._wcs_tolerance,
+                                **kwargs)
+
+        spectrum._spectral_unit = spectral_unit
+
+        return spectrum
+
+    def with_beam(self, beam):
+        '''
+        Attach a new beam object to the OneDSpectrum.
+
+        Parameters
+        ----------
+        beam : `~radio_beam.Beam`
+            A new beam object.
+        '''
+
+        meta = self.meta.copy()
+        meta['beam'] = beam
+
+        return self._new_spectrum_with(beam=beam, meta=meta)
+
+
+class VaryingResolutionOneDSpectrum(BaseOneDSpectrum, MultiBeamMixinClass):
+
+    def __new__(cls, value, beams=None, read_beam=False, **kwargs):
+        self = super(VaryingResolutionOneDSpectrum, cls).__new__(cls, value, **kwargs)
+
         if beams is None:
-            if hasattr(self, 'beams'):
-                kwargs['beams'] = beams
-        elif beam is None:
-            if hasattr(self, 'beam'):
-                kwargs['beam'] = beam
+            if "beams" in self.meta:
+                beams = self.meta['beams']
+            elif read_beam:
+                beams = cube_utils.try_load_beams(self.header)
+                if beams is None:
+                    warnings.warn("Cannot load beams table from header.",
+                                  BeamWarning
+                                  )
 
         if beams is not None:
-            cls = VaryingResolutionOneDSpectrum
-        else:
-            cls = OneDSpectrum
-            
+            self.beams = beams
+            self.meta['beams'] = beams
+
+        return self
+
+    def with_beams(self, beams):
+        '''
+        Attach a new beams object to the VaryingResolutionOneDSpectrum.
+
+        Parameters
+        ----------
+        beams : `~radio_beam.Beams`
+            A new beams object.
+        '''
+
+        meta = self.meta.copy()
+        meta['beams'] = beams
+
+        return self._new_spectrum_with(beams=beams, meta=meta)
+
+    @property
+    def hdu(self):
+        warnings.warn("There are multiple beams for this spectrum that "
+                      "are being ignored when creating the HDU.",
+                      BeamWarning
+                     )
+        return super(VaryingResolutionOneDSpectrum, self).hdu
+
+    @property
+    def hdulist(self):
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            hdu = self.hdu
+
+        beamhdu = cube_utils.beams_to_bintable(self.beams)
+
+        return HDUList([hdu, beamhdu])
+
+    def _new_spectrum_with(self, data=None, wcs=None, mask=None, meta=None,
+                           fill_value=None, spectral_unit=None, unit=None,
+                           header=None, wcs_tolerance=None, beams=None,
+                           **kwargs):
```

Comment:
isn't most of this redundant?  Could we use `super()` here instead of rewriting all this code?  So we could have `BaseOneDSpectrum._new_spectrum_with()` that would take care of everything except the beams, and have the `beam` or `beams` handling be in the subclasses?

Reply:
Fixed.